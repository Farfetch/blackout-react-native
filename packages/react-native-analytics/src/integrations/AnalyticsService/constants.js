export const BACKGROUND_APP_STATE = 'background';
export const INACTIVE_APP_STATE = 'inactive';
