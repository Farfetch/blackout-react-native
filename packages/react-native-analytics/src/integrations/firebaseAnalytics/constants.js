export const MAX_PRODUCT_CATEGORIES = 5;

export const OPTION_EVENTS_MAPPER = 'eventsMapper';
export const OPTION_SCREEN_VIEWS_MAPPER = 'screenViewsMapper';
export const OPTION_ON_SET_USER_HANDLER = 'onSetUser';
export const OPTION_SET_CUSTOM_USER_ID_PROPERTY = 'setCustomUserIdProperty';
export const OPTION_GOOGLE_CONSENT_CONFIG = 'googleConsentConfig';

export const MESSAGE_PREFIX = '[FirebaseAnalytics]';
