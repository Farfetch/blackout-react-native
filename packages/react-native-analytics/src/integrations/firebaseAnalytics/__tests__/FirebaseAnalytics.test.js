import FirebaseAnalyticsIntegration from '../';
import { eventsMapper } from '../mapper';
import {
  eventTypes,
  integrations,
  trackTypes as analyticsTrackTypes,
  utils,
} from '@farfetch/blackout-core/analytics';
import firebaseAnalytics from '@react-native-firebase/analytics';
import { LOGIN_METHOD } from '../constants';

const mockFirebaseAnalyticsReturn = {
  logScreenView: jest.fn(),
  logLogin: jest.fn(),
  setUserId: jest.fn(),
  setUserProperties: jest.fn(),
  logEvent: jest.fn(),
  logSelectItem: jest.fn(),
};

jest.mock('@react-native-firebase/analytics', () => {
  return {
    __esModule: true,
    default: jest.fn(() => mockFirebaseAnalyticsReturn),
  };
});

jest.mock('@farfetch/blackout-core/analytics', () => ({
  ...jest.requireActual('@farfetch/blackout-core/analytics'),
  utils: {
    logger: {
      error: jest.fn(),
    },
  },
}));

function createInstance(options, loadData) {
  const instance = FirebaseAnalyticsIntegration.createInstance(
    options,
    loadData,
  );

  return instance;
}

beforeEach(() => {
  jest.clearAllMocks();
});

describe('FirebaseAnalyticsIntegration integration', () => {
  it('Should extend the abstract class `Integration`', () => {
    expect(FirebaseAnalyticsIntegration.prototype).toBeInstanceOf(
      integrations.Integration,
    );
  });

  it('`shouldLoad` should return false if there is no user consent', () => {
    expect(FirebaseAnalyticsIntegration.shouldLoad({ statistics: false })).toBe(
      false,
    );
    expect(FirebaseAnalyticsIntegration.shouldLoad()).toBe(false);
    expect(FirebaseAnalyticsIntegration.shouldLoad({})).toBe(false);
  });

  it('`shouldLoad` should return true if there is user consent', () => {
    expect(FirebaseAnalyticsIntegration.shouldLoad({ statistics: true })).toBe(
      true,
    );
  });

  it('Should return a FirebaseAnalyticsIntegration instance from createInstance', () => {
    expect(FirebaseAnalyticsIntegration.createInstance({})).toBeInstanceOf(
      FirebaseAnalyticsIntegration,
    );
  });

  it('Should throw an error if "@react-native-firebase/analytics" is not installed', () => {
    jest.resetModules();

    jest.doMock('@react-native-firebase/analytics', () => undefined);

    const FreshFirebaseAnalyticsIntegration = require('..').default;

    expect(() => FreshFirebaseAnalyticsIntegration.createInstance()).toThrow(
      '[FirebaseAnalytics]: "@react-native-firebase/analytics" package is not installed. Please, make sure you have this dependency installed before using this integration.',
    );
  });
});

describe('FirebaseAnalyticsIntegration instance', () => {
  it('Should allow to extend the eventsMapper', () => {
    const customEventsMapper = {
      myEvent: () => 'myEvent',
    };

    const instance = createInstance({ eventsMapper: customEventsMapper });

    expect(instance.eventsMapper).toEqual({
      ...eventsMapper,
      ...customEventsMapper,
    });
  });

  describe('User', () => {
    it('Should call `onSetUser` when the instance is created', () => {
      const spy = jest.spyOn(
        FirebaseAnalyticsIntegration.prototype,
        'onSetUser',
      );
      const loadData = {
        user: {
          id: 123,
          traits: {},
        },
      };

      createInstance(null, loadData);

      expect(spy).toHaveBeenCalledWith(loadData, null);
    });

    it('Should call custom `onSetUser` if passed via the integration options', () => {
      const customOnSetUserSpy = jest.fn();
      const loadData = {
        foo: 'bar',
        biz: 1,
      };
      const options = {
        onSetUser: 'foo',
      };
      createInstance(options, loadData);

      expect(utils.logger.error).toHaveBeenCalledWith(
        '[FirebaseAnalytics] TypeError: "onSetUser" is not a function. If you are passing a custom "onSetUser" property to the integration, make sure you are passing a valid function.',
      );

      options.onSetUser = customOnSetUserSpy;

      createInstance(options, loadData);

      expect(customOnSetUserSpy).toHaveBeenCalledWith(loadData);
    });

    it('Should call firebase analytics methods when `onSetUser` is called', async () => {
      const userData = {
        user: {
          id: 123,
          traits: {
            email: 'asdasd@asdasd.com',
            name: 'asd',
            isGuest: false,
          },
        },
      };
      const instance = createInstance();

      await instance.onSetUser(userData);

      expect(firebaseAnalytics().logLogin).toHaveBeenCalledWith({
        method: LOGIN_METHOD,
      });
      expect(firebaseAnalytics().setUserId).toHaveBeenCalledWith(
        userData.user.id.toString(),
      );
      expect(firebaseAnalytics().setUserProperties).toHaveBeenCalledWith({
        ...userData.user.traits,
        isGuest: userData.user.traits.isGuest.toString(),
      });

      firebaseAnalytics().logLogin.mockClear();
      firebaseAnalytics().setUserId.mockClear();
      firebaseAnalytics().setUserProperties.mockClear();

      // Test when a user is guest - should not call any method
      await instance.onSetUser({ id: 123, traits: { isGuest: true } });

      expect(firebaseAnalytics().logLogin).not.toHaveBeenCalled();
      expect(firebaseAnalytics().setUserId).not.toHaveBeenCalled();
      expect(firebaseAnalytics().setUserProperties).not.toHaveBeenCalled();

      // Force onSetUser with the same id for the next test
      await instance.onSetUser(userData);

      firebaseAnalytics().logLogin.mockClear();
      firebaseAnalytics().setUserId.mockClear();
      firebaseAnalytics().setUserProperties.mockClear();

      // Force onSetUser with the same id, it should not call the methods again
      await instance.onSetUser(userData);

      expect(firebaseAnalytics().logLogin).not.toHaveBeenCalled();
      expect(firebaseAnalytics().setUserId).not.toHaveBeenCalled();
      expect(firebaseAnalytics().setUserProperties).not.toHaveBeenCalled();

      // Test logout
      await instance.onSetUser({
        user: {
          id: 12321123123123,
          traits: {
            isGuest: true,
          },
        },
      });

      expect(firebaseAnalytics().setUserId).toHaveBeenCalledWith(null);
      expect(firebaseAnalytics().setUserProperties).toHaveBeenCalledWith({});
    });
  });

  describe('track', () => {
    it('Should track a screen', async () => {
      const trackScreenData = {
        type: analyticsTrackTypes.SCREEN,
        event: 'foo',
      };
      const instance = createInstance();

      await instance.track(trackScreenData);

      expect(firebaseAnalytics().logScreenView).toHaveBeenCalledWith({
        screen_class: 'foo',
        screen_name: 'foo',
      });
    });

    it('Should track an event', async () => {
      const analyticsTrackEventData = {
        type: analyticsTrackTypes.TRACK,
        event: eventTypes.PRODUCT_CLICKED,
        properties: {
          foo: 'bar',
        },
      };
      const validFirebaseAnalyticsMethod = 'logSelectItem';
      let instance = createInstance();

      // Test a non mapped event - It should not be tracked
      await instance.track({
        ...analyticsTrackEventData,
        event: 'myCustomEventThatShouldBeIgnored',
      });

      expect(firebaseAnalytics().logEvent).not.toHaveBeenCalled();

      // Test a mapped event that do not return a valid function
      instance = createInstance({
        eventsMapper: {
          [eventTypes.PRODUCT_CLICKED]: 'foo',
        },
      });

      await instance.track(analyticsTrackEventData);

      expect(utils.logger.error).toHaveBeenCalledWith(
        '[FirebaseAnalytics] TypeError: Event mapping for event "Product Clicked" is not a function. If you\'re passing a custom event mapping for this event, make sure a function is passed.',
      );

      utils.logger.error.mockClear();

      // Test a mapped event that returns a valid function with invalid `properties`
      instance = createInstance({
        eventsMapper: {
          [eventTypes.PRODUCT_CLICKED]: () => ({
            method: 'logSelectItem',
            properties: 'foo',
          }),
        },
      });

      await instance.track(analyticsTrackEventData);

      expect(utils.logger.error).toHaveBeenCalledWith(
        '[FirebaseAnalytics] TypeError: The properties passed for event Product Clicked is not an object. If you are passing a custom event mapping for this event, make sure you return a valid object under "properties" key.',
      );

      utils.logger.error.mockClear();

      // Test a mapped event that returns a valid function with valid `properties`
      instance = createInstance({
        eventsMapper: {
          [eventTypes.PRODUCT_CLICKED]: () => ({
            method: validFirebaseAnalyticsMethod,
            properties: analyticsTrackEventData.properties,
          }),
        },
      });

      await instance.track(analyticsTrackEventData);

      expect(
        firebaseAnalytics()[validFirebaseAnalyticsMethod],
      ).toHaveBeenCalledWith(analyticsTrackEventData.properties);

      // Test a mapped event that returns a valid function with valid `properties` without a known method for firebase
      const unknownMethodOfFirebase = 'unknownMethodOfFirebase';
      instance = createInstance({
        eventsMapper: {
          [eventTypes.PRODUCT_CLICKED]: () => ({
            method: unknownMethodOfFirebase,
            properties: analyticsTrackEventData.properties,
          }),
        },
      });

      await instance.track(analyticsTrackEventData);

      expect(utils.logger.error).toHaveBeenCalledWith(
        '[FirebaseAnalytics] Method "unknownMethodOfFirebase" is not defined. If you are passing a custom event mapping, make sure you return a supported Firebase Analytics event.',
      );

      firebaseAnalytics().logEvent.mockClear();
      // Test a mapped event that returns a valid function with valid `properties` without a high-level method
      const customEvent = 'my custom event';
      const formattedCustomEvent = 'my_custom_event';

      instance = createInstance({
        eventsMapper: {
          [customEvent]: () => ({
            method: undefined,
            properties: analyticsTrackEventData.properties,
          }),
        },
      });

      await instance.track({ ...analyticsTrackEventData, event: customEvent });

      expect(firebaseAnalytics().logEvent).toHaveBeenCalledWith(
        formattedCustomEvent,
        analyticsTrackEventData.properties,
      );
    });
  });
});
