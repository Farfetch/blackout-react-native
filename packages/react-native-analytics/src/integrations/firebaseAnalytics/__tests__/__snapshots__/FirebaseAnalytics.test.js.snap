// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Newsletter events Should track multiple gender event 1`] = `
Array [
  Array [
    "sign_up_newsletter",
    Object {
      "newsletter_gender": "Woman,Man",
    },
  ],
  Array [
    "sign_up_newsletter",
    Object {
      "newsletter_gender": "WOMAN,MAN",
    },
  ],
  Array [
    "sign_up_newsletter",
    Object {
      "newsletter_gender": "Woman,Man",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Newsletter events Should track single gender event 1`] = `
Array [
  Array [
    "sign_up_newsletter",
    Object {
      "newsletter_gender": "Woman",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Search events Should track search event search term instead of search query. 1`] = `
Array [
  "search",
  Object {
    "search_term": "term",
  },
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Search events Should track search event search without search term but with search query instead. 1`] = `
Array [
  "search",
  Object {
    "search_term": "query",
  },
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Address Info Added event correctly 1`] = `
Array [
  Array [
    "add_address_info",
    Object {
      "address_finder": undefined,
      "coupon": "ACME2019",
      "currency": "USD",
      "delivery_type": undefined,
      "packaging_type": undefined,
      "shipping_tier": "Next Day",
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the App Opened event correctly 1`] = `
Array [
  Array [
    "app_open",
    undefined,
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Checkout Abandoned event correctly 1`] = `
Array [
  Array [
    "abandon_confirmation_checkout",
    Object {
      "coupon": "ACME2019",
      "currency": "USD",
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Checkout Started event correctly 1`] = `
Array [
  Array [
    "begin_checkout",
    Object {
      "coupon": "ACME2019",
      "currency": "USD",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_list_id": undefined,
          "item_list_name": undefined,
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": 1,
          "size": "L",
        },
      ],
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Checkout Step Editing event correctly 1`] = `
Array [
  Array [
    "edit_checkout_step",
    Object {
      "checkout_step": 1,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Filters Applied event correctly 1`] = `
Array [
  Array [
    "apply_filters",
    Object {
      "filters": "{\\"brands\\":[2765,4062],\\"categories\\":[135973],\\"colors\\":[1],\\"discount\\":[0],\\"gender\\":[0],\\"price\\":[0,1950],\\"sizes\\":[16]}",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Filters Cleared event correctly 1`] = `
Array [
  Array [
    "clear_filters",
    Object {
      "filters": "{\\"brands\\":[2765,4062],\\"categories\\":[135973],\\"colors\\":[1],\\"discount\\":[0],\\"gender\\":[0],\\"price\\":[0,1950],\\"sizes\\":[16]}",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Interact Content event correctly 1`] = `
Array [
  Array [
    "interact_content",
    Object {
      "content_type": "biz",
      "interaction_type": "Click",
      "some_other_property": 12312312,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Login event correctly 1`] = `
Array [
  Array [
    "login",
    Object {
      "method": "Acme",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Order Completed event correctly 1`] = `
Array [
  Array [
    "purchase",
    Object {
      "affiliation": undefined,
      "coupon": "ACME2019",
      "currency": "USD",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_list_id": undefined,
          "item_list_name": undefined,
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": 1,
          "size": "L",
        },
      ],
      "shipping": 3.6,
      "tax": 2.04,
      "transaction_id": "50314b8e9bcf000000000000",
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Order Refunded event correctly 1`] = `
Array [
  Array [
    "refund",
    Object {
      "affiliation": undefined,
      "coupon": undefined,
      "currency": "USD",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_list_id": undefined,
          "item_list_name": undefined,
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": 1,
          "size": "L",
        },
      ],
      "shipping": undefined,
      "tax": undefined,
      "transaction_id": "50314b8e9bcf000000000000",
      "value": 19,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Payment Info Added event correctly 1`] = `
Array [
  Array [
    "add_payment_info",
    Object {
      "coupon": "ACME2019",
      "currency": "USD",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_list_id": undefined,
          "item_list_name": undefined,
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": 1,
          "size": "L",
        },
      ],
      "payment_type": "credit card",
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Place Order Started event correctly 1`] = `
Array [
  Array [
    "place_order",
    Object {
      "affiliation": undefined,
      "coupon": "ACME2019",
      "currency": "USD",
      "shipping": 3.6,
      "tax": 2.04,
      "transaction_id": "50314b8e9bcf000000000000",
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Product Added to Cart event correctly 1`] = `
Array [
  Array [
    "add_to_cart",
    Object {
      "currency": "USD",
      "from": "Wishlist",
      "item_list_id": "d3618128-5aa9-4caa-a452-1dd1377a6190",
      "item_list_name": "my_wishlist",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": 1,
          "size": "L",
        },
      ],
      "value": 19,
      "wishlist_id": undefined,
      "wishlist_name": undefined,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Product Added to Wishlist event correctly 1`] = `
Array [
  Array [
    "add_to_wishlist",
    Object {
      "currency": "USD",
      "from": "PLP",
      "item_list_id": "/en-pt/shopping/woman",
      "item_list_name": "Woman shopping",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": undefined,
          "size": undefined,
        },
      ],
      "value": 19,
      "wishlist_id": "d3618128-5aa9-4caa-a452-1dd1377a6190",
      "wishlist_name": undefined,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Product Clicked event correctly 1`] = `
Array [
  Array [
    "select_item",
    Object {
      "from": "PLP",
      "item_list_id": "/en-pt/shopping/woman",
      "item_list_name": "Woman shopping",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "GBP",
          "discount": 6,
          "index": 3,
          "item_brand": undefined,
          "item_id": "507f1f77bcf86cd799439011",
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": undefined,
          "location_id": undefined,
          "price": 25,
          "quantity": undefined,
          "size": undefined,
        },
      ],
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Product List Viewed event correctly 1`] = `
Array [
  Array [
    "view_item_list",
    Object {
      "error": undefined,
      "filters": undefined,
      "from": "PLP",
      "item_list_id": undefined,
      "item_list_name": "Woman shopping",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": 2,
          "item_brand": undefined,
          "item_id": "507f1f77bcf86cd799439011",
          "item_list_id": "09a35590-bb62-4027-a630-5da04ec64fb5",
          "item_list_name": "Woman shopping",
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": undefined,
          "location_id": undefined,
          "price": 25,
          "quantity": undefined,
          "size": undefined,
        },
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": 3,
          "item_brand": undefined,
          "item_id": "507f1f77bcf86cd799439012",
          "item_list_id": "09a35590-bb62-4027-a630-5da04ec64fb5",
          "item_list_name": "Woman shopping",
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": undefined,
          "location_id": undefined,
          "price": 25,
          "quantity": undefined,
          "size": undefined,
        },
      ],
      "sort_option": undefined,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Product Removed From Wishlist event correctly 1`] = `
Array [
  Array [
    "remove_from_wishlist",
    Object {
      "affiliation": undefined,
      "coupon": undefined,
      "currency": "USD",
      "discount": 6,
      "from": "PLP",
      "index": undefined,
      "item_brand": "Just A T-Shirt",
      "item_category": "Clothing",
      "item_category2": "Tops",
      "item_category3": "T-shirts",
      "item_id": "507f1f77bcf86cd799439011",
      "item_list_id": "/en-pt/shopping/woman",
      "item_list_name": "Woman shopping",
      "item_name": "Gareth McConnell Dreamscape T-Shirt",
      "item_variant": "Black",
      "location_id": undefined,
      "price": 25,
      "quantity": undefined,
      "size": undefined,
      "value": 19,
      "wishlist_id": "d3618128-5aa9-4caa-a452-1dd1377a6190",
      "wishlist_name": undefined,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Product Removed from Cart event correctly 1`] = `
Array [
  Array [
    "remove_from_cart",
    Object {
      "currency": "USD",
      "from": "Bag",
      "item_list_id": "e0030b3c-b970-4496-bc72-f9a38d6270b1",
      "item_list_name": "Bag",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": 1,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": 1,
          "size": "L",
        },
      ],
      "value": 19,
      "wishlist_id": undefined,
      "wishlist_name": undefined,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Product Viewed event correctly 1`] = `
Array [
  Array [
    "view_item",
    Object {
      "currency": "USD",
      "from": "PLP",
      "image_count": undefined,
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": undefined,
          "size": undefined,
        },
      ],
      "value": 19,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Promocode Applied event correctly 1`] = `
Array [
  Array [
    "apply_promo_code",
    Object {
      "address_finder": undefined,
      "coupon": "ACME2019",
      "currency": "USD",
      "delivery_type": undefined,
      "packaging_type": undefined,
      "shipping_tier": "Next Day",
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Select Content event correctly 1`] = `
Array [
  Array [
    "select_content",
    Object {
      "content_type": "biz",
      "item_id": 12312312,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Share event correctly 1`] = `
Array [
  Array [
    "share",
    Object {
      "content_type": "image",
      "item_id": "123456",
      "method": "Facebook",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Shipping Info Added event correctly 1`] = `
Array [
  Array [
    "add_shipping_info",
    Object {
      "address_finder": undefined,
      "coupon": "ACME2019",
      "currency": "USD",
      "delivery_type": undefined,
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": "USD",
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_list_id": undefined,
          "item_list_name": undefined,
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": 1,
          "size": "L",
        },
      ],
      "packaging_type": undefined,
      "shipping_tier": "Next Day",
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Shipping Method Added event correctly 1`] = `
Array [
  Array [
    "add_shipping_method",
    Object {
      "address_finder": undefined,
      "coupon": "ACME2019",
      "currency": "USD",
      "delivery_type": undefined,
      "packaging_type": undefined,
      "shipping_tier": "Next Day",
      "value": 24.64,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Sign-up Form Completed event correctly 1`] = `
Array [
  Array [
    "sign_up",
    Object {
      "method": "Acme",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the Sign-up Newsletter event correctly 1`] = `
Array [
  Array [
    "sign_up_newsletter",
    Object {
      "newsletter_gender": "Woman",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the bag event correctly 1`] = `
Array [
  Array [
    "view_cart",
    Object {
      "currency": "USD",
      "from": "Bag",
      "item_list_id": "e0030b3c-b970-4496-bc72-f9a38d6270b1",
      "item_list_name": "Bag",
      "items": Array [
        Object {
          "affiliation": undefined,
          "coupon": undefined,
          "currency": undefined,
          "discount": 6,
          "index": undefined,
          "item_brand": "Just A T-Shirt",
          "item_category": "Clothing",
          "item_category2": "Tops",
          "item_category3": "T-shirts",
          "item_id": "507f1f77bcf86cd799439011",
          "item_list_id": undefined,
          "item_list_name": undefined,
          "item_name": "Gareth McConnell Dreamscape T-Shirt",
          "item_variant": "Black",
          "location_id": undefined,
          "price": 25,
          "quantity": 1,
          "size": "L",
        },
      ],
      "value": 19,
      "wishlist_id": undefined,
      "wishlist_name": undefined,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the bag event correctly 2`] = `
Array [
  Array [
    Object {
      "currency": "USD",
      "from": "Bag",
      "list": "Bag",
      "list_id": "e0030b3c-b970-4496-bc72-f9a38d6270b1",
      "screen_class": "bag",
      "screen_name": "bag",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the search event correctly 1`] = `
Array [
  Array [
    "search",
    Object {
      "search_term": "shoes",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the search event correctly 2`] = `
Array [
  Array [
    Object {
      "currency": "EUR",
      "screen_class": "search",
      "screen_name": "search",
      "search_query": "shoes",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the wishlist event correctly 1`] = `
Array [
  Array [
    "view_wishlist",
    Object {
      "wishlist_id": "d3618128-5aa9-4caa-a452-1dd1377a6190",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Should map the wishlist event correctly 2`] = `
Array [
  Array [
    Object {
      "currency": "USD",
      "from": "Wishlist",
      "screen_class": "wishlist",
      "screen_name": "wishlist",
      "wishlist_id": "d3618128-5aa9-4caa-a452-1dd1377a6190",
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Update Product events Should send "change_colour" event to firebase 1`] = `
Array [
  Array [
    "change_colour",
    Object {
      "colour": "red",
      "from": "Bag",
      "item_id": "507f1f77bcf86cd799439011",
      "item_name": "Gareth McConnell Dreamscape T-Shirt",
      "old_colour": undefined,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Update Product events Should send "change_quantity" event to firebase 1`] = `
Array [
  Array [
    "change_quantity",
    Object {
      "from": "Bag",
      "item_id": "507f1f77bcf86cd799439011",
      "item_name": "Gareth McConnell Dreamscape T-Shirt",
      "old_quantity": undefined,
      "quantity": 1,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Update Product events Should send "change_quantity", "change_size" and "change_colour" events to firebase when a product updated event contains all the necessary changes 1`] = `
Array [
  Array [
    "change_quantity",
    Object {
      "from": "Bag",
      "item_id": "507f1f77bcf86cd799439011",
      "item_name": "Gareth McConnell Dreamscape T-Shirt",
      "old_quantity": undefined,
      "quantity": 1,
    },
  ],
  Array [
    "change_size",
    Object {
      "from": "Bag",
      "item_id": "507f1f77bcf86cd799439011",
      "item_name": "Gareth McConnell Dreamscape T-Shirt",
      "old_size": undefined,
      "size": "L",
    },
  ],
  Array [
    "change_colour",
    Object {
      "colour": "red",
      "from": "Bag",
      "item_id": "507f1f77bcf86cd799439011",
      "item_name": "Gareth McConnell Dreamscape T-Shirt",
      "old_colour": undefined,
    },
  ],
]
`;

exports[`FirebaseAnalyticsIntegration instance Track Default events mappings Update Product events Should send "change_size" event to firebase 1`] = `
Array [
  Array [
    "change_quantity",
    Object {
      "from": "Bag",
      "item_id": "507f1f77bcf86cd799439011",
      "item_name": "Gareth McConnell Dreamscape T-Shirt",
      "old_quantity": undefined,
      "quantity": 1,
    },
  ],
  Array [
    "change_colour",
    Object {
      "colour": "red",
      "from": "Bag",
      "item_id": "507f1f77bcf86cd799439011",
      "item_name": "Gareth McConnell Dreamscape T-Shirt",
      "old_colour": undefined,
    },
  ],
]
`;
